validator/sawtooth_validator/concurrent/atomic.py

```diff
@@ -46,3 +47,80 @@ def inc(self, step=1):
     def dec(self, step=1):
         with self._lock:
             self._value -= step
+
+
+class Set:
+    def __init__(self):
+        self.set = set()
+        self.lock = RLock()
+
+    def add(self, element):
+        with self.lock:
+            self.set.add(element)
+
+    def remove(self, element):
+        with self.lock:
+            self.set.remove(element)
+
+    def __contains__(self, element):
+        with self.lock:
+            return element in self.set
+
+
+class MapOfList:
```

Comment:
This is often called a `MultiMap`.  Given that it has a lock, maybe even `ConcurrentMultiMap`